go/vt/orchestrator/logic/topology_recovery.go

```diff
@@ -815,113 +613,120 @@ func checkAndRecoverDeadPrimary(analysisEntry inst.ReplicationAnalysis, candidat
+		if err != nil {
+			return false, nil, err
+		}
+		candidateTabletAlias = candidateTablet.Alias
+	}
 	topologyRecovery, err = AttemptRecoveryRegistration(&analysisEntry, !forceInstanceRecovery, !forceInstanceRecovery)
 	if topologyRecovery == nil {
 		AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("found an active or recent recovery on %+v. Will not issue another RecoverDeadPrimary.", analysisEntry.AnalyzedInstanceKey))
 		return false, nil, err
 	}
 	log.Infof("Analysis: %v, deadprimary %+v", analysisEntry.Analysis, analysisEntry.AnalyzedInstanceKey)
 
-	// That's it! We must do recovery!
-	// TODO(sougou): This function gets called by GracefulPrimaryTakeover which may
-	// need to obtain shard lock before getting here.
-	unlock, err := LockShard(analysisEntry.AnalyzedInstanceKey)
-	if err != nil {
-		log.Infof("CheckAndRecover: Analysis: %+v, InstanceKey: %+v, candidateInstanceKey: %+v, "+
-			"skipProcesses: %v: NOT detecting/recovering host, could not obtain shard lock (%v)",
-			analysisEntry.Analysis, analysisEntry.AnalyzedInstanceKey, candidateInstanceKey, skipProcesses, err)
-		return false, nil, err
-	}
-	defer unlock(&err)
-
-	// Check if someone else fixed the problem.
-	tablet, err := TabletRefresh(analysisEntry.AnalyzedInstanceKey)
-	if err == nil && tablet.Type != topodatapb.TabletType_PRIMARY {
-		// TODO(sougou); use a version that only refreshes the current shard.
-		RefreshTablets()
+	// this check is needed because sometimes DeadPrimary code path is forcefully spawned off from other recoveries like PrimaryHasPrimary.
+	// So we need to check that we only run an ERS if the instance that we analyzed was actually a primary! Otherwise, we would end up running an ERS
+	// even when the cluster is fine or the problem can be fixed via some other recovery
+	if tablet.Type != topodatapb.TabletType_PRIMARY {
+		RefreshTablets(true)
 		AuditTopologyRecovery(topologyRecovery, "another agent seems to have fixed the problem")
-		// TODO(sougou): see if we have to reset the cluster as healthy.
 		return false, topologyRecovery, nil
 	}
 
-	AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("will handle DeadPrimary event on %+v", analysisEntry.ClusterDetails.ClusterName))
-	recoverDeadPrimaryCounter.Inc(1)
-	recoveryAttempted, promotedReplica, lostReplicas, err := recoverDeadPrimary(topologyRecovery, candidateInstanceKey, skipProcesses)
-	if err != nil {
-		AuditTopologyRecovery(topologyRecovery, err.Error())
+	// check if we have received SIGTERM, if we have, we should not continue with the recovery
+	val := atomic.LoadInt32(&hasReceivedSIGTERM)
+	if val > 0 {
+		return false, topologyRecovery, errors.New("Can't lock shard: SIGTERM received")
 	}
-	topologyRecovery.LostReplicas.AddInstances(lostReplicas)
-	if !recoveryAttempted {
-		return false, topologyRecovery, err
+
+	// check if we have received an ERS in progress, if we do, we should not continue with the recovery
+	if checkAndSetIfERSInProgress() {
```

Comment:
Another vtorc instance will not interfere with ERS since the first step in ERS is to lock the shard and only 1 ERS instance will be able to hold the lock at any given time. However, this point has made me realise there might be a separate issue, which is of reloading the ephemeral data. When 1 vtorc instance runs an ERS, it is going to change the primary tablet and the replication sources of all the others. Unless the other vtorc instances reload their ephemeral data in their backend, they are going to be using the old data to run other failover scenarios. In the case of vtorc which ran the ERS, this is prevented by using `RefreshTablets(true /* forceRefresh */)` which forcefully refreshes all the ephemeral information. I think we need to think about this in greater detail too. I am linking @deepthi  and @sougou  for their thoughts on this too. However, I think that this is a problem out of scope of this PR, but we should get around to it before a GA release of vtorc.

Reply:
I guess I'm confused then why this is needed in the first place.  The lock should guarantee that only one ERS is in progress. Why we need to add this protection?

Reply:
We need this for now, because LockShard is blocking. It only guarantees that only 1 instance will be running an ERS. But it could happen that an ERS is in progress and another call to ERS is issued which would be blocked. When the first ERS succeeds, the next ERS would get unblocked and start its execution (if its lock shard timeour hasn't expired). Eventually, in vtorc we would want to first lock the shard and then refresh all the ephemeral information and then check once again if the failover is required. If it is, then we go ahead with it. This change however would be across all the failover scenarios of vtorc, so I haven't done it as part of this PR but we will. When we do, we can remove this check too.

Reply:
I agree this looks confusing. With plain orchestrator, only the "leader" node runs recoveries. Now, it's possible for the leader node to run a recovery, and then lose leadership, in which case a new leader steps in. So how do we prevent the new leader from immediately running an unwanted recovery (it may have a stale impression of the state)? Any node assuming leadership must go through a "cool down" period. It will not run recoveries before 3*polling interval seconds have passed; this allows the newly elected leader to get a good grasp of the situation before it makes new decisions.

In `vtorc` design this does not happen; any `vtorc` node can run recoveries, and they compete over `LockShard`. Moreover, they don't share recovery info, which is yet another thing `orchestrator` does. So here I'm thinking we need to discuss the design and how to prevent near-immediate followups to recoveries.